.grit/patterns/no_self_assigned_variables.md

```diff
@@ -0,0 +1,64 @@
+---
+title: Variables should not be self-assigned
+---
+
+# Variables should not be self-assigned
+
+It is redundant and usually a bug when a variable is assigned to itself.
+
+tags: #java
+
+```grit
+language java
+
+assignment_expression($left, $right) as $assignment where {
+    $left <: `$right`,
+    or {
+        and {
+            $left <: not contains `this`,
+            $left => `this.$left`,
+        },
+        $assignment <: within expression_statement() as $exp where {
+            $exp => .,
+        }
+    },
+}
+```
+
+## Optimistically rewrites variable to instance variable when it is not one
+
+```java
+class Watermelon {
+    public void setName(String name) {
+        name = name;
+    }
+}
+```
+
+```java
+class Watermelon {
+    public void setName(String name) {
```

Comment:
I believe this is not correct Java syntax, doesn't the instance variable need to be declared? Wherever possible we want "corrected" code to be valid.

Reply:
I don't think I want to impose a restriction or add a declaration because it could be inherited from a parent class. We could just remove it in all cases. It might be incorrect more often but it would be guaranteed to be syntactically valid.

Reply:
Right, I don't mean a pattern change - just a change in the example. We don't want examples to look "obviously incorrect."